core/trino-main/src/main/java/io/trino/execution/scheduler/NodeRequirements.java

```diff
@@ -28,11 +29,13 @@
 {
     private final Optional<CatalogName> catalogName;
     private final Set<HostAddress> addresses;
+    private final Optional<DataSize> memory;
```

Comment:
It could be changed so this one is obligatory. Yet currently the `TaskSource` is producing `NodeRequirements` without this field set. We only attach memory requirements in `SqlQueryScheduler` via `PartitionMemoryEstimator`.
One can argue that maybe it should happen earlier, as tasksource may have knowledge about the nature of processing to be done, hence about memory requirements. On the other hand  the state of a cluster (shape of nodes, and utilization) also may impact how much memory we want to assign for tasks. If cluster is highly utilized we may be more conservative, and risk more task failures, than if we have plenty of space. 

I would not change it right now, and improve as a followup when we know more.

Reply:
Memory required by a task is technically a product of the size of shuffle inputs + splits assigned to a give tasks. The more splits and inputs we assign the more default memory should be given to a task so it is with some probability is likely to succeed (can be derived empirically or based on history).

Currently there are session properties that control these values:

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/SystemSessionProperties.java#L151

For example if the cluster is configured to create tasks that contain at most 5GB of shuffle inputs and 20 splits (256MB each on average) the memory requirement can be conservatively configured to, say 10GB to ensure that the task is "likely" to succeed with the configured inputs. 

How much memory is available on a node is more of a proxy to the concurrency limit, e.g.: "how many tasks can be run in parallel on a node with this amount of memory available". We may also allow over-commit based on empirical observation.

Reply:
Yeah - I think our opinion do not differ here. I wanted to still be able to use different memory limit for a task, than a limit estimated by task source. But it is possible with current abstractions. PTAL.